### DISCUSSION TITLE
Polly's built-in OnHedging metric in resilience.polly.strategy.events doesn't match with custom metrics we record in onHedging action callback

### DISCUSSION THREAD

**Discussion by lvxiaoxin** [App-vNext/Polly#2354] [Q&A] [answered]
Hi Polly experts,

We recently adopt Polly resilience and chaos through the Microsoft.Extensions.Http.Resilience package (version 8.3), and applied the _AddStandardHedgingHandler_ to our HttpClient. As a result, we automatically get the Polly built-in telemetry like `resilience.polly.strategy.events`. We start to do some chaos testing experiment, however, we noticed that the metrics count of `event.name == "OnHedging" ` doesn't match metric we records in onHedging action while setting up the resilience handlers for HttpClient. 

In our experiment, we used the `_AddStandardHedgingHandler_` to our HttpClient and configured it with primary/secondary endpoint group. Then we introduced a chaos fault injection with failure ratio as 0.8. We could see the circuit breaker strategy kick in and take effects, then the hedging strategy works and switches the target endpoint from primary to secondary. Everything works well except for the `resilience.polly.strategy.events` metrics.

![image](https://github.com/user-attachments/assets/f6c97e90-baf8-4854-a875-d3b41c4b1137)

The chart is aggregated per 1 minute, and it shows in average there are only 6 OnHedging metrics, while actually there are about 2k+ requests per minute. 

Then we did an experiment that we manually send a custom metric in the `onHedging` action callback while setting up the resilience policy, and its count match with the request traffic, but doesn't match above `resilience.polly.strategy.events` metrics.
![image](https://github.com/user-attachments/assets/275dd562-344e-4bf1-ad21-b2013f9376ff)

So, I believe the hedging operation indeed happens and the functionality indeed works, but the only problem is unexpected `resilience.polly.strategy.events` that it's having too less metrics filtering by `event.name == "OnHedging".

Does my understanding of the built-in metrics wrong? Shouldn't it match with the onHedging action?

**Comment by martincostello** [comment]
@martintmk I feel you're well placed to answer this?

**Comment by martintmk** [comment, marked as answer]
If we look at this code: 

https://github.com/App-vNext/Polly/blob/6f7edf8913b2af07c1d929c8fdfe9bdb67314efe/src/Polly.Core/Hedging/Controller/TaskExecution.cs#L139

Notice, that path where OnHedging is called and where the telemetry is reported, is identical.

These metrics should be identical. Maybe you are collecting these differently?

**Comment by lvxiaoxin** [comment]
Thanks @martintmk and @martincostello. Yes you are right. We originally looked into the `HandleOnHedgingAsync` method and considered they should match from the code logic, but saw unexpected results from the final collected metrics. 

It turns out that the difference comes from the way we collecting those metrics (one is sampling as count and another is not), leading to the mis-match between the built-in metric and my custom record metric. After I fixed that, the OnHedging metric of `resilience.polly.strategy.events` start to make senses and match our experiment traffic.

Thanks a lot for the help here!
